function  WWHBookData_AddTOCEntries(P)
{
var A=P.fN("Welcome to the REMBRANDT Online Help","0");
A=P.fN("Online Help Tips","1");
var B=A.fN("Using REMBRANDT Online Help","1#1063568");
A=P.fN("About REMBRANDT 1.5","2");
B=A.fN("About REMBRANDT","3");
B=A.fN("About REMBRANDT Functions","4");
A=P.fN("Getting Started with REM\u00adBRANDT 1.5","5");
B=A.fN("Creating a User Account","6");
B=A.fN("Logging In","7");
B=A.fN("Accepting REMBRANDT Provisions","8");
B=A.fN("Welcome to REMBRANDT","9");
B=A.fN("Getting Help","10");
B=A.fN("Logging Out","11");
A=P.fN("Conducting a Simple Search","12");
B=A.fN("Simple Search Overview","13");
B=A.fN("Gene Expression Simple Search","14");
var C=B.fN("Understanding a Gene Expression Plot","15");
var D=C.fN("Geometric Mean Plot Details","16");
D=C.fN("Log2 Intensity Gene Expression Plot Details","17");
D=C.fN("Box and Whisker Log2 Intensity Gene Expression Plot Details","18");
var E=D.fN("Displaying a Coin Plot","19");
D=C.fN("Saving, Printing, and E-mailing a Gene Expression Plot","20");
B=A.fN("K-M Gene Expression Simple Search","21");
C=B.fN("Eliminating Aliases","22");
C=B.fN("Redrawing the K-M Survival Plot for Gene Expression Data","23");
C=B.fN("Understanding K-M Survival Plot for Gene Expression Data","24");
B=A.fN("K-M Copy Number Simple Search","25");
C=B.fN("Redrawing the K-M Survival Plot for Copy Number Data","26");
C=B.fN("Understanding K-M Survival Plot for Copy Number Data","27");
B=A.fN("K-M Sample Simple Search","28");
C=B.fN("Understanding K-M Survival Plot for Sample Data","29");
B=A.fN("Viewing the Clinical Reports","30");
B=A.fN("Viewing Clinical Plots","31");
A=P.fN("Conducting Advanced Searches","32");
B=A.fN("Advanced Searches Overview","33");
B=A.fN("Gene Expression Advanced Search","34");
C=B.fN("Selecting a Gene Ontology (GO) Classification","35");
C=B.fN("Selecting a Pathway","36");
B=A.fN("Copy Number Advanced Search","37");
B=A.fN("Clinical Study Advanced Search","38");
B=A.fN("Managing Advanced Searches","39");
B=A.fN("Refining a Query","40");
A=P.fN("High Order Analysis","41");
B=A.fN("High Order Analysis Overview","42");
B=A.fN("Performing a Class Comparison","43");
B=A.fN("Performing a Principal Component Analysis","44");
B=A.fN("Performing Hierarchical Clustering Analysis","45");
A=P.fN("Viewing Results","46");
B=A.fN("Results Overview","47");
B=A.fN("Advanced Search or Query Results","48");
C=B.fN("Gene Expression Sample Report","49");
D=C.fN("Filtering Results by Gene or Reporter (Filter Toolbar)","50");
D=C.fN("Highlighting Results By Value (Highlight Toolbar)","51");
D=C.fN("Selecting and Saving Sample Results (Select Samples Toolbar)","52");
D=C.fN("Differentiating Data (Show All Values Toolbar)","53");
D=C.fN("Removing Columns (Hide Diseases toolbar)","54");
D=C.fN("Showing Additional Information","55");
C=B.fN("Gene Expression Disease Report","56");
C=B.fN("Copy Number Sample Report","57");
D=C.fN("Displaying Copy Number Sample Data in webGenome","58");
B=A.fN("High Order Analysis Results","59");
C=B.fN("Class Comparison Report","60");
D=C.fN("Filtering a p-value (Filter p-value toolbar)","61");
D=C.fN("Selecting and Saving Reporters (Select Reporters toolbar)","62");
D=C.fN("Resorting Column Results","63");
C=B.fN("Principal Component Analysis Plot","64");
C=B.fN("Hierarchical Clustering Report","65");
B=A.fN("Downloading BRB Files","66");
A=P.fN("Managing Lists","67");
B=A.fN("Managing Lists Overview","68");
B=A.fN("Viewing the Data Items in a List","69");
B=A.fN("Removing Data Items to Create a New List","70");
B=A.fN("Combining Existing Lists to Create a New List","71");
B=A.fN("Deleting an Entire List","72");
B=A.fN("Adding a New \u201cCustom\u201d List","73");
A=P.fN("Glossary","74");
}
