function  WWHBookData_AddTOCEntries(P)
{
var A=P.fN("Welcome to the REMBRANDT 1.5 Online Help","0");
A=P.fN("Online Help Tips","1");
var B=A.fN("Using REMBRANDT Online Help","1#1063568");
A=P.fN("About REMBRANDT 1.5","2");
B=A.fN("About REMBRANDT","3");
B=A.fN("About REMBRANDT Functions","4");
A=P.fN("Getting Started with REM\u00adBRANDT 1.5","5");
B=A.fN("Creating a User Account","6");
B=A.fN("Logging In","7");
B=A.fN("Accepting REMBRANDT Provisions","8");
B=A.fN("REMBRANDT 1.5 Workspace","9");
B=A.fN("Getting Help","10");
B=A.fN("Application Support","11");
B=A.fN("Logging Out","12");
A=P.fN("Conducting a Simple Search","13");
B=A.fN("Simple Search Overview","14");
B=A.fN("Gene Expression Simple Search","15");
var C=B.fN("Eliminating Aliases","16");
C=B.fN("Understanding a Gene Expression Plot","17");
var D=C.fN("Geometric Mean Plot Details","18");
D=C.fN("Log2 Intensity Gene Expression Plot Details","19");
D=C.fN("Box and Whisker Log2 Intensity Gene Expression Plot Details","20");
var E=D.fN("Displaying a Coin Plot","21");
D=C.fN("Saving, Printing, and E-mailing a Gene Expression Plot","22");
B=A.fN("K-M Gene Expression Simple Search","23");
C=B.fN("Redrawing the K-M Survival Plot for Gene Expression Data","24");
C=B.fN("Understanding K-M Survival Plot for Gene Expression Data","25");
B=A.fN("K-M Copy Number Simple Search","26");
C=B.fN("Redrawing the K-M Survival Plot for Copy Number Data","27");
C=B.fN("Understanding K-M Survival Plot for Copy Number Data","28");
B=A.fN("K-M Sample Search","29");
C=B.fN("Understanding K-M Survival Plot for Sample Data","30");
B=A.fN("Viewing the Clinical Reports","31");
B=A.fN("Viewing Clinical Plots","32");
A=P.fN("Conducting Advanced Searches","33");
B=A.fN("Advanced Searches Overview","34");
B=A.fN("Gene Expression Advanced Search","35");
C=B.fN("Selecting a Gene Ontology (GO) Classification","36");
C=B.fN("Selecting a Pathway","37");
B=A.fN("Copy Number Advanced Search","38");
B=A.fN("Clinical Study Advanced Search","39");
B=A.fN("Managing Advanced Searches","40");
B=A.fN("Refining a Query","41");
A=P.fN("High Order Analysis","42");
B=A.fN("High Order Analysis Overview","43");
B=A.fN("Performing a Class Comparison","44");
B=A.fN("Performing a Principal Component Analysis","45");
B=A.fN("Performing Hierarchical Clustering Analysis","46");
A=P.fN("Viewing Results","47");
B=A.fN("Results Overview","48");
B=A.fN("Advanced Search or Query Results","49");
C=B.fN("Gene Expression Sample Report","50");
D=C.fN("Filtering Results by Gene or Reporter (Filter Toolbar)","51");
D=C.fN("Highlighting Results By Value (Highlight Toolbar)","52");
D=C.fN("Selecting and Saving Sample Results (Select Samples Toolbar)","53");
D=C.fN("Differentiating Data (Show All Values Toolbar)","54");
D=C.fN("Removing Columns (Hide Diseases toolbar)","55");
D=C.fN("Showing Additional Information","56");
C=B.fN("Gene Expression Disease Report","57");
C=B.fN("Copy Number Sample Report","58");
D=C.fN("Displaying Copy Number Sample Data in webGenome","59");
B=A.fN("High Order Analysis Results","60");
C=B.fN("Class Comparison Report","61");
D=C.fN("Filtering a p-value (Filter p-value Toolbar)","62");
D=C.fN("Selecting and Saving Reporters (Select Reporters toolbar)","63");
D=C.fN("Resorting Column Results","64");
C=B.fN("Principal Component Analysis Plot","65");
C=B.fN("Hierarchical Clustering Report","66");
B=A.fN("Downloading BRB Array Tools and Files","67");
A=P.fN("Managing Lists","68");
B=A.fN("Managing Lists Overview","69");
B=A.fN("Viewing the Data Items in a List","70");
B=A.fN("Removing Data Items to Create a New List","71");
B=A.fN("Combining Existing Lists to Create a New List","72");
B=A.fN("Deleting an Entire List","73");
B=A.fN("Adding a New \u201cCustom\u201d List","74");
A=P.fN("Glossary","75");
}
