function  WWHBookData_AddTOCEntries(P)
{
var A=P.fN("Welcome to the REMBRANDT Online Help","0");
A=P.fN("Online Help Tips","1");
var B=A.fN("Using REMBRANDT Online Help","1#1063568");
A=P.fN("About REMBRANDT 1.5","2");
B=A.fN("About REMBRANDT","3");
B=A.fN("About REMBRANDT Functions","4");
A=P.fN("Getting Started with REM\u00adBRANDT 1.5","5");
B=A.fN("Creating a User Account","6");
B=A.fN("Logging In","7");
B=A.fN("Accepting REMBRANDT Provisions","8");
B=A.fN("Welcome to REMBRANDT","9");
B=A.fN("Getting Help","10");
B=A.fN("Logging Out","11");
A=P.fN("Conducting a Simple Search","12");
B=A.fN("Simple Search Overview","13");
B=A.fN("Gene Expression Simple Search","14");
var C=B.fN("Understanding a Gene Expression Plot","15");
var D=C.fN("Show Log2 Gene Expression Plot Details","16");
D=C.fN("Show Raw Gene Expression Plot Details","17");
D=C.fN("Show BW Gene Expression Plot Details","18");
D=C.fN("Saving, Printing, and E-mailing a Gene Expression Plot","19");
B=A.fN("K-M Gene Expression Simple Search","20");
C=B.fN("Eliminating Aliases","21");
C=B.fN("Redrawing the K-M Survival Plot for Gene Expression Data","22");
C=B.fN("Understanding K-M Survival Plot for Gene Expression Data","23");
B=A.fN("K-M Copy Number Simple Search","24");
C=B.fN("Redrawing the K-M Survival Plot for Copy Number Data","25");
C=B.fN("Understanding K-M Survival Plot for Copy Number Data","26");
B=A.fN("K-M Sample Simple Search","27");
C=B.fN("Understanding K-M Survival Plot for Sample Data","28");
B=A.fN("Viewing the Clinical Reports","29");
B=A.fN("Viewing Clinical Plots","30");
A=P.fN("Conducting Advanced Searches","31");
B=A.fN("Advanced Searches Overview","32");
B=A.fN("Gene Expression Advanced Search","33");
C=B.fN("Selecting a Gene Ontology (GO) Classification","34");
C=B.fN("Selecting a Pathway","35");
B=A.fN("Copy Number Advanced Search","36");
B=A.fN("Clinical Study Advanced Search","37");
B=A.fN("Managing Advanced Searches","38");
B=A.fN("Refining a Query","39");
A=P.fN("High Order Analysis","40");
B=A.fN("High Order Analysis Overview","41");
B=A.fN("Performing a Class Comparison","42");
B=A.fN("Performing a Principal Component Analysis","43");
B=A.fN("Performing Hierarchical Clustering Analysis","44");
A=P.fN("Viewing Results","45");
B=A.fN("Results Overview","46");
B=A.fN("Advanced Search or Query Results","47");
C=B.fN("Gene Expression Sample Report","48");
D=C.fN("Filtering Results by Gene or Reporter (Filter Toolbar)","49");
D=C.fN("Highlighting Results By Value (Highlight Toolbar)","50");
D=C.fN("Selecting and Saving Sample Results (Select Samples Toolbar)","51");
D=C.fN("Differentiating Data (Show All Values Toolbar)","52");
D=C.fN("Removing Columns (Hide Diseases toolbar)","53");
D=C.fN("Showing Additional Information","54");
C=B.fN("Gene Expression Disease Report","55");
C=B.fN("Copy Number Sample Report","56");
D=C.fN("Displaying Copy Number Sample Data in webGenome","57");
B=A.fN("High Order Analysis Results","58");
C=B.fN("Class Comparison Report","59");
D=C.fN("Filtering a p-value (Filter p-value toolbar)","60");
D=C.fN("Selecting and Saving Reporters (Select Reporters toolbar)","61");
D=C.fN("Resorting Column Results","62");
C=B.fN("Principal Component Analysis Plot","63");
C=B.fN("Viewing Clinical Plots","64");
C=B.fN("Hierarchical Clustering Report","65");
A=P.fN("Managing Lists","66");
B=A.fN("Managing Lists Overview","67");
B=A.fN("View List Details","68");
B=A.fN("Creating Custom Lists by Adding and Deleting Entire Lists or Data from Lists","69");
C=B.fN("Deleting Items From a List","69#1080563");
C=B.fN("Deleting Entire Lists","70");
C=B.fN("Adding a New List","71");
B=A.fN("Creating New Custom Lists from Existing Lists","72");
A=P.fN("Glossary","73");
}
